### ISSUE TITLE
Neon.Temporal status

### ISSUE THREAD

**Issue by jishf** [nforgeio/neonKUBE#1012]
Hey team - I hope you're all well!

I got in touch back in April with @jefflill over email about Cadence/Temporal and your .NET libraries. We (Ionburst) have been looking at integrating Temporal into our stack for a while - and as a .NET Core shop, we're interesting in using Neon.Temporal.

We've been keeping an eye on progress until we had time to start exploring Temporal fully. We're now in that place, so I was hoping to find out if Neon.Temporal is ready to use? I appreciate it's still under active development and we're not looking for a production-ready release, just something we can start to play around with.

Alongside that, we're more than happy to be guinea pigs and help with bug testing and scale/load-testing further down the line.

Cheers! 😄

**Comment by jefflill** [comment]
**HI:** We're all doing fine and I hope you are too,

We're excited to see Temporal v1 launch.  We started the Temporal port a few months ago but sidelined it for several weeks while Temporal was making their final breaking changes and one of started a new job.

We're planning on getting started back into this beginning next week.  As you probably remember, the current Cadence implementation is actually based on the Cadence GO client and a wrapper program we developed that uses HTTP to proxy calls from .NET our GO proxy and then on to the Cadence server.  We're going to continue with this strategy for Temporal in the near term, but we hope to build our proxy as a DLL/shared library and drop the HTTP stuff and just marshal data between .NET and GO in memory buffers.  We experimented this briefly a year ago and has some problems which we now believe were unrelated.  This will be quite a bit nicer.

**Neon.Cadence** has been in serious test/production for a year or so now by a sister company and we really haven't seen much trouble (although traffic is pretty low).  These services are deployed as .NET Core to Linux containers running in Kubernetes.  We've seen 2-3 cases where the .NET side lost connectivity with the GO proxy and we had to restart the containers (there's a way to detect this situation from .NET).  We hope this will be moot after we convert to DLLs.

It's too early to give you an ETA on this.  We'll need to get back into this for a few week.  I'm crossing my fingers that perhaps we'll see something beginning to work in 4-6 weeks.

There are a couple important gaps right now in Cadence and Temporal being a port, has the same issues:

* There's currently no way for a workflow to wait for one of multiple things to happen, like waiting for one of two child workflows to complete or a signal to be received.  We believe we've thought this through.  This will take a bit of refactoring.

* We also haven't doing anything around workflow unit testing support yet.

Our plan is to:

1. Publish one or more alpha releases in maybe 4-6 weeks when most things work again (excluding the multi-wait and unit testing)
2. After we've added the multi-wait and unit testing and it looks reasonable, we'll start doing preview releases and we'll port our sister companies production cluster over.
3. Our thinking is that we'll continue doing preview releases into 2021 that are still based on the GO proxy but expected to be usable for production.
4. Q1 2021, we're planning on beginning a native gRPC implementation and dropping the GO proxy completely,  I haven't scoped how much work this will be (but hope springs eternal, so hopefully it won't be too bad).  Once this is complete and it looks good we'll make this a production release.

I expect the native gRPC production release to work much the same as he GO proxied one.  I think the main changes will be that some of the API methods will no longer be `async` and I think we'll be able to replace some of the workflow/activity **_futures_** related classes with standard .NET `Task` semantics.  We ended with more async APIs than are really necessary due to the additional layer of async-ness due to the HTTP communication with the GO proxy.

We really do welcome your interest and would love to have more eyes to be sure we get it right.  I've been super excited about this stuff since I saw Maxim do a Cadence talk 3-4 years ago.  I think the .NET community will love it.

--Jeff    jeff@lilltek.com 

@johncburns1 
@marcusbooyah

**Comment by jishf** [comment]
Fantastic, thanks for the update!

**Comment by surumadurum** [comment]
Hi there, how has this situation developed? Best, Robert

**Comment by johncburns1** [comment]
@surumadurum thanks for checking in! Naturally our development slowed down a bit due to unforeseeable circumstances (COVID).   We are still actively working on Neon.Temporal and it is nearing stability.  

At the moment we are wrapping up Exceptions and Exception handling.  Once Exception handling is complete, Neon.Temporal will be usable featuring much of the same functionality Neon.Cadence provides.  After this initial port is complete we plan to implement the features mentioned [here](https://github.com/nforgeio/neonKUBE/issues/1012#issuecomment-706487379) by @jefflill.

Thanks for the patience and if you have any more specific questions or just want an update please do not hesitate to reach out!

**Comment by jefflill** [comment]
This has been on the back burner but we just decided to restart the effort because the current Cadence/Temporal implementations have stopped working on the latest .NET SDKs due to ASP.NET runtime conflicts.

We currently deploy a GOLANG proxy process that communicates with Cadence/Temporal servers and then bridge calls from the C# side to the proxy via HTTP.  The new plan is to convert the proxies to shared libraries (DLLs on Windows) so we'll be able to just do in-process calls.  We has this close to working a couple years ago but we had to move on to other priorities,

We're going to track this effort here: #1295

The Temporal implementation is otherwise fairly far along.  @johncburns1 has some error handling work remaining and I believe there were still a few unit test failures.
